sql/core/src/main/scala/org/apache/spark/sql/ui/SQLListener.scala

```diff
@@ -0,0 +1,340 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.ui
+
+import scala.collection.mutable
+
+import org.apache.spark.{AccumulatorParam, JobExecutionStatus}
+import org.apache.spark.executor.TaskMetrics
+import org.apache.spark.scheduler._
+import org.apache.spark.sql.{DataFrame, SQLContext}
+import org.apache.spark.sql.execution.SQLExecution
+
+private[sql] class SQLListener(sqlContext: SQLContext) extends SparkListener {
+
+  private val retainedExecutions =
+    sqlContext.sparkContext.conf.getInt("spark.sql.ui.retainedExecutions", 1000)
+
+  private val activeExecutions = mutable.HashMap[Long, SQLExecutionUIData]()
+
+  // Old data in the following fields must be removed in "trimExecutionsIfNecessary".
+  // If adding new fields, make sure "trimExecutionsIfNecessary" can clean up old data
+
+  // VisibleForTesting
+  val executionIdToData = mutable.HashMap[Long, SQLExecutionUIData]()
```

Comment:
we should probably make this fully `private val _executionIdToData`, and then have a public getter `def executionIdToData: Map[Long, SQLExecutionUIData]`. I would recommend that we do it for other data structures in this class too instead of just exposing the size, so we can assert something stronger.